### ISSUE TITLE
Scheduler and cluster auto scaler don't agree on available resources

### ISSUE THREAD

**Issue by dennis-benzinger-hybris** [kubernetes/kubernetes#93186]
**What happened**:
Sometimes we see a pod not being scheduled because the scheduler thinks there are not enough resources:
```
I0713 16:28:36.138069       1 filter_out_schedulable.go:60] Filtering out schedulables
I0713 16:28:36.138724       1 filter_out_schedulable.go:106] Pod default.backoffice-57687f7994-4frt4 marked as unschedulable can be scheduled on node k8s-pr213821ktte-17230105-vmss000000. Ignoring in scale up.
I0713 16:28:36.138748       1 filter_out_schedulable.go:116] 1 pods marked as unschedulable can be scheduled.
I0713 16:28:36.138779       1 filter_out_schedulable.go:74] Schedulable pods present
I0713 16:28:36.138801       1 static_autoscaler.go:389] No unschedulable pods
```
But on the other hand the cluster auto scaler thinks there are enough resources:
```
I0713 16:27:11.266164       1 factory.go:462] Unable to schedule default/backoffice-57687f7994-4frt4: no fit: 0/4 nodes are available: 1 Insufficient cpu, 2 Insufficient memory.; waiting
I0713 16:28:41.267187       1 factory.go:462] Unable to schedule default/backoffice-57687f7994-4frt4: no fit: 0/4 nodes are available: 1 Insufficient cpu, 2 Insufficient memory.; waiting
I0713 16:30:11.267529       1 factory.go:462] Unable to schedule default/backoffice-57687f7994-4frt4: no fit: 0/4 nodes are available: 1 Insufficient cpu, 2 Insufficient memory.; waiting
```
So the pod doesn't get scheduled but also no new worker node is being created.

**What you expected to happen**:
Either the pod is scheduled or the cluster auto scaler adds a worker VM.

**How to reproduce it (as minimally and precisely as possible)**:
Unfortunately we don't have an easy way to reproduce it yet. It might be related to big differences between resource requests and limit. If possible I'll add instructions to reproduce later.

**Anything else we need to know?**:

**Environment**:
- Kubernetes version (use `kubectl version`): 1.18.2
- Cloud provider or hardware configuration: Azure
- OS (e.g: `cat /etc/os-release`): Ubuntu 18.04.4 LTS
- Kernel (e.g. `uname -a`): 5.0.0-1036-azure
- Install tools: ARM template generated by [aks-engine v0.50.0](https://github.com/Azure/aks-engine/releases/tag/v0.50.0)
- Network plugin and version (if this is a network-related bug): not network related
- Others:
Cluster auto scaler version is 1.18.1 and its deployment is based on https://github.com/Azure/aks-engine/blob/master/parts/k8s/addons/cluster-autoscaler.yaml

**Comment by ahg-g** [comment]
do you use init containers? if so, it might be related to https://github.com/kubernetes/autoscaler/issues/3224

**Comment by MaciekPytel** [comment]
/sig autoscaling

This looks like it may be init containers problem linked by @ahg-g. That one will be fixed in Cluster Autoscaler release 1.18.2, which is scheduled to be cut tomorrow. 
If you're not using init containers can you provide some more information about your config? In particular any parts of pod spec that you could share would be helpful.

**Comment by dennis-benzinger-hybris** [comment]
We use init containers but not on the pod which wasn't scheduled. Can this issue affect other pods which don't have init containers themselves?

kubernetes/autoscaler#3224 mentions that the issue "happens if you have pods whose initContainer requests are bigger than their main container requests". Does that include the case where the init containers don't specify any resources but the regular containers do?

Here's the unschedulable pod as returned by K8s (minus some confidential details): 
```yaml
apiVersion: v1
kind: Pod
metadata:
  creationTimestamp: "2020-07-13T15:46:55Z"
  generateName: backoffice-57687f7994-
  labels:
    pod-template-hash: 57687f7994
  name: backoffice-57687f7994-4frt4
  namespace: default
  resourceVersion: "42653"
  selfLink: /api/v1/namespaces/default/pods/backoffice-57687f7994-4frt4
  uid: ddedfab9-8a03-47ff-b93b-557cc84ac08a
spec:
  containers:
  - args:
    - backoffice
    imagePullPolicy: IfNotPresent
    lifecycle:
      preStop:
        exec:
          command:
          - /bin/sh
          - -c
          - echo preStop
    livenessProbe:
      failureThreshold: 3
      httpGet:
        httpHeaders:
        - name: X-Forwarded-Proto
          value: https
        path: /liveness
        port: tomcat-http
        scheme: HTTP
      initialDelaySeconds: 1800
      periodSeconds: 30
      successThreshold: 1
      timeoutSeconds: 2
    name: platform
    ports:
    - containerPort: 8081
      name: tomcat-http
      protocol: TCP
    - containerPort: 8088
      name: tomcat-internal
      protocol: TCP
    readinessProbe:
      failureThreshold: 1
      httpGet:
        httpHeaders:
        - name: X-Forwarded-Proto
          value: https
        path: /readiness
        port: tomcat-http
        scheme: HTTP
      initialDelaySeconds: 45
      periodSeconds: 30
      successThreshold: 1
      timeoutSeconds: 2
    resources:
      limits:
        cpu: "2"
        memory: 6400Mi
      requests:
        cpu: "2"
        memory: 6400Mi
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  imagePullSecrets:
  - name: image-registry-secret
  priority: 0
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  terminationGracePeriodSeconds: 300
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2020-07-13T15:46:55Z"
    message: '0/4 nodes are available: 1 Insufficient cpu, 2 Insufficient memory.'
    reason: Unschedulable
    status: "False"
    type: PodScheduled
  phase: Pending
  qosClass: Guaranteed
```

**Comment by ahg-g** [comment]
Yes, it will affect other pods because it impacts the total resources presumed to be allocated on the node.

**Comment by ahg-g** [comment]
can we close this as duplicate of kubernetes/autoscaler#3224?

**Comment by dennis-benzinger-hybris** [comment]
I would prefer if we could keep it open for a few more days. The symptom aligns very nicely with kubernetes/autoscaler#3224 but as that issue explains "pods whose initContainer requests are bigger than their main container requests" are affected. And I haven't found such pods in our setup.

**Comment by dennis-benzinger-hybris** [comment]
Finally I could reproduce the issue and have a running cluster in that state. Here are some files which show the situation:

This shows all nodes with their allocatable resources and the allocated resources:
[allocated-resources.txt](https://github.com/kubernetes/kubernetes/files/4959075/allocated-resources.txt)

Log of the cluster autoscaler:
[logs.cluster-autoscaler.redacted.log](https://github.com/kubernetes/kubernetes/files/4959076/logs.cluster-autoscaler.redacted.log)

Log of the scheduler:
[logs.scheduler.redacted.log](https://github.com/kubernetes/kubernetes/files/4959077/logs.scheduler.redacted.log)

The pod which isn't scheduled:
[pod.backoffice.redacted.json.txt](https://github.com/kubernetes/kubernetes/files/4959087/pod.backoffice.redacted.json.txt)

All pods which have initcontainers and their resources:
[pods.with-initcontainers.json.txt](https://github.com/kubernetes/kubernetes/files/4959088/pods.with-initcontainers.json.txt)

I didn't find any pod whose initcontainers request more resources than the main containers.

More logs can be uploaded if necessary.

**Comment by dennis-benzinger-hybris** [comment]
Do you have any hints on how to debug this further? Are there other logs I should upload?

**Comment by MaciekPytel** [comment]
There is no pod with 'backoffice' in the name in CA logs. That's very unexpected if a pod is pending. In fact there is almost nothing in this log - is this a result of redacting? What --v value are you using (CA logs are generally only useful with --v=4)?

**Comment by dennis-benzinger-hybris** [comment]
The recent logs are with `--v=3`. The logs in the original description are from a different, by now decommissioned cluster.

I'll try to reproduce with `--v=4`.

**Comment by MaciekPytel** [comment]
May be worth waiting for new cluster autoscaler release before trying that. I've run into a bunch of problems building it, but I'm mostly done with them and I should be able to cut the patch release on Monday.

**Comment by dennis-benzinger-hybris** [comment]
So far we couldn't reproduce the issue with Kubernetes v1.18.8. We think it was fixed by #93387.
